fw/access_log.c

```diff
@@ -266,6 +277,123 @@ process_truncated(TfwStr *in, BasicStr *out, char *p, char *end,
 	}
 }
 
+static void
+do_access_log_req_mmap(TfwHttpReq *req, u16 resp_status,
+		       u32 resp_content_length)
+{
+	u64 *dropped = this_cpu_ptr(&mmap_log_dropped);
+	TfwBinLogEvent *event;
+	unsigned int room_size;
+	TfwStr referer, ua;
+	u32 resp_time;
+	char *data, *p;
+
+	if (!mmap_buffer)
+		goto drop;
+
+	room_size = tfw_mmap_buffer_get_room(mmap_buffer, &data);
+	if (room_size < sizeof(TfwBinLogEvent))
+		goto drop;
+
+	event = (TfwBinLogEvent *)data;
+	p = data + sizeof(TfwBinLogEvent);
+
+#define WRITE_TO_BUF(val, size)			\
+	do {					\
+		if (room_size < size)		\
+			goto drop;		\
+		memcpy_fast(p, val, size);	\
+		p += size;			\
+		room_size -= size;		\
+	} while (0)
+
+#define WRITE_FIELD(field, val)							\
+	do {									\
+		if (TFW_MMAP_LOG_FIELD_IS_SET(event, TFW_MMAP_LOG_##field)) {	\
+			WRITE_TO_BUF(&val, sizeof(val));			\
+		}								\
+	} while (0)
+
+	event->timestamp = jiffies;
+
+	if (*dropped && room_size >= sizeof(u64)) {
+		event->type = TFW_MMAP_LOG_TYPE_DROPPED;
+
+		memcpy_fast(p, dropped, sizeof(u64));
+		tfw_mmap_buffer_commit(mmap_buffer, p - data);
```

Comment:
It seems that `tfw_mmap_buffer_commit` can fails (very little chance here). But we should not zeroed dropped in this case and update room_size and p

Reply:
Now this call placed only at the end of the function.
Added a warning and `goto drop` for tfw_mmap_buffer_commit fails.